File: solutions/1126.py
n = int(input())

blocks = list(map(int, input().split()))

cache = {}

def solution(current_index, height_difference):
    if (current_index, height_difference) in cache:
        return cache[(current_index, height_difference)]

    if current_index == len(blocks):
        if height_difference == 0:
            return 0
        else:
            return -1000000

    a = solution(current_index + 1, height_difference)
    b = blocks[current_index] + solution(current_index + 1, height_difference - blocks[current_index]);
    c = solution(current_index + 1, height_difference + blocks[current_index]);

    cache[(current_index, height_difference)] = max(a, b, c)
    return max(a, b, c)

t = solution(0, 0)
if t <= 0:
    print(-1)
else:
    print(t)

File: solutions/1126.cpp
#include <iostream>
#include <limits>
#include <vector>
#include "../modules/Input.h"
#include "../modules/FastIO.h"
#include "../modules/Matrix.h"

std::vector<int> blocks;

Matrix<int> cache(50, 500'000, std::numeric_limits<int>::min());

int solution(size_t index, int delta) {
	if (index == blocks.size()) {
		if (delta == 0) return 0;
		return std::numeric_limits<int>::min();
	}

	int& result = cache(index, delta);

	if (result == std::numeric_limits<int>::min()) {
		result = std::max(result, solution(index + 1, delta));
		result = std::max(result, solution(index + 1, delta + blocks[index]));
		if (delta < blocks[index])
			result = std::max(result, delta + solution(index + 1, blocks[index] - delta));
		else
		 	result = std::max(result, blocks[index] + solution(index + 1, delta - blocks[index]));
	}
	return result;
}

int main() {
	FastIO();
	int n;
	std::cin >> n;
	blocks = Input::toVec<int>(n);

	int result = solution(0, 0);
	if (result <= 0) std::cout << "-1";
	else std::cout << result;
}
